### ISSUE TITLE
Negative Gaussian Log Likelihood loss function implementation

### ISSUE THREAD

**Issue by IvanMM27** [giogiopg/graphnet#8]
These loss functions that have been actively used for training models  with OrcaNeT framework in KM3NeT. They have been succesfully implemented in GraphNeT framework. The advantage of this loss function is that is suitable for estimating a error of the reconstruction as well as its prediction. This turns really useful when applying quality cuts on the reconstructed variables in the same way is done for classical reconstruction methods, based on the likelihood of the reconstruction.

In order to add them, the `graphnet/src/graphnet/training/loss_functions.py` script must be modified accordingly.

```
"""Collection of loss functions.

All loss functions inherit from `LossFunction` which ensures a common syntax,
handles per-event weights, etc.
"""

from abc import abstractmethod
from typing import Any, Optional, Union, List, Dict

import numpy as np
import scipy.special
import torch
from torch import Tensor
from torch import nn
from torch.nn.functional import (
    one_hot,
    cross_entropy,
    binary_cross_entropy,
    softplus,
    gaussian_nll_loss,
)

from graphnet.models.model import Model
from graphnet.utilities.decorators import final

...

class GaussianNegativeLogLikelihood1D(LossFunction):
    """Gaussian Negative Log Likelihood loss function for mean and variance estimation in a single dimension."""

    def __init__(
        self,
        full: bool = True,
        eps: float = 1e-7,
        log10: bool = True,
        *args: Any,
        **kwargs: Any,
    ) -> None:
        self.full = full
        self.eps = eps
        self.log10 = log10
        super().__init__(*args, **kwargs)
    
    def _forward(self, prediction: Tensor, target: Tensor) -> Tensor:
        """Calculate Gaussian Negative Log Likelihood loss for a given mean and variance estimation in a single dimension.
        
        Args:
            prediction: Output of the model. Must have shape [N, 2] where 0th
                column is a prediction of `mean` and 1st column is an estimate
                of `variance`.
            target: Target tensor, extracted from graph object.

        Returns:
            loss: Elementwise Gaussian Negative Log Likelihood loss terms. Shape [N,]
        """
        # Check(s)
        assert prediction.dim() == 2 and prediction.size()[1] == 2
        assert target.dim() == 2
        assert prediction.size()[0] == target.size()[0]

        # Formatting target
        estimation_true = target.float()
        
        # Formatting prediction
        if self.log10:
            mean_pred = torch.log10(prediction[:, 0])
        else:
            mean_pred = prediction[:, 0]
        variance_pred = softplus(prediction[:, 1])
               
        return gaussian_nll_loss(mean_pred, estimation_true, variance_pred, full = self.full, eps = self.eps, reduction='none')
    
class GaussianNegativeLogLikelihood3D(LossFunction):
    """Gaussian Negative Log Likelihood loss function for mean and variance estimation in three dimensions."""

    def __init__(
        self,
        full: bool = True,
        eps: float = 1e-7,
        *args: Any,
        **kwargs: Any,
    ) -> None:
        self.full = full
        self.eps = eps
        super().__init__(*args, **kwargs)
    
    def _forward(self, prediction: Tensor, target: Tensor) -> Tensor:
        """Calculate Gaussian Negative Log Likelihood loss for a given mean and variance estimation in three dimensions.
        
        Args:
            prediction: Output of the model. Must have shape [N, 6] where the 
                first 3 columns are predictions of `mean` and second 3 columns
                are estimations of `variance`.
            target: Target tensor, extracted from graph object.

        Returns:
            loss: Elementwise Gaussian Negative Log Likelihood loss terms. Shape [N,]
        """
        # Check(s)
        assert prediction.dim() == 2 and prediction.size()[1] == 6
        assert target.dim() == 2 and target.size()[1] == 3
        assert prediction.size()[0] == target.size()[0]

        # Formatting target
        estimation_true = target.float()
        
        # Formatting prediction
        mean_pred = prediction[:, :3]
        variance_pred = softplus(prediction[:, 3:])
        
        return gaussian_nll_loss(mean_pred, estimation_true, variance_pred, full = self.full, eps = self.eps, reduction='none')
```

In order to use them, below an example is provided
```
# For direction reconstruction, assuming we reconstruct the individual coordinates x, y an z
from graphnet.models.task.reconstruction import DirectionReconstructionWithUncertainty
from graphnet.training.loss_functions import GaussianNegativeLogLikelihood3D

task = DirectionReconstructionWithUncertainty(
                                                                                              hidden_size = backbone.nb_outputs, 
                                                                                              target_labels = config["task"]["target"],
                                                                                              loss_function = GaussianNegativeLogLikelihood3D(),
    )
    
# For energy reconstruction
from graphnet.models.task.reconstruction import EnergyReconstructionWithUncertainty
from graphnet.training.loss_functions import GaussianNegativeLogLikelihood1D

task = EnergyReconstructionWithUncertainty(
                                                    hidden_size = backbone.nb_outputs, 
                                                    target_labels = config["task"]["target"],
                                                    loss_function = GaussianNegativeLogLikelihood1D(log10 = True),
                                                    transform_target = lambda x: torch.log10(x),
                                                    transform_inference = lambda x: torch.stack([torch.pow(10, x[:, 0]), x[:, 1]], dim = 1)
)
```
